source/extensions/filters/http/local_ratelimit/local_ratelimit.cc

```diff
@@ -115,13 +158,80 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
   return Http::FilterHeadersStatus::StopIteration;
 }
 
+Http::FilterHeadersStatus Filter::encodeHeaders(Http::ResponseHeaderMap& headers, bool) {
+  const auto* config = getConfig();
+
+  if (config->enableXRateLimitHeaders()) {
+    auto descriptors = decoder_callbacks_->streamInfo()
+                           .filterState()
+                           ->getDataMutable<LocalRateLimitRequestDescriptorsQueue>(
+                               LocalRateLimitRequestDescriptorsQueue::key())
+                           .pop();
+    if (descriptors.has_value()) {
+      auto limit = maxTokens(descriptors.value());
+      auto remaining = remainingTokens(descriptors.value());
+      auto reset = remainingFillInterval(descriptors.value());
+
+      headers.addCopy(HttpFilters::Common::RateLimit::XRateLimitHeaders::get().XRateLimitLimit,
+                      limit);
+      headers.addCopy(HttpFilters::Common::RateLimit::XRateLimitHeaders::get().XRateLimitRemaining,
+                      remaining);
+      headers.addCopy(HttpFilters::Common::RateLimit::XRateLimitHeaders::get().XRateLimitReset,
+                      reset);
+    }
```

Comment:
Is this trying to retrieve the descriptors added in the queue during decodeHeaders? I'm not sure if this would do the right thing for multiplexed connections (e.g. H2), since there is no guarantee that the responses are processed in the same order as the requests.

Could we just store the descriptors on the filter itself? That would guarantee that the descriptors used in encode* is the same as those decided on in decode*

Or maybe I'm misunderstanding and we're using a queue for some other reason?

Reply:
Yes, we must store necessary information (descriptor, which is populated by the request headers) for `x-ratelimit-*` headers in `decodeHeaders`, and will generate `x-ratelimit-*` headers into the response headers in `encodeHeaders`. So here I store the descriptor in the stream info since we no longer have access to request headers when we are in `encodeHeaders`.

So you mentioned in multiplexed connections, these connections will share a single stream info and may not enter the `encodeHeaders` in order. Is it correct? If so, then there will be a mistake.

Reply:
It would be reused betweeen connections since you're storing it with Connection lifetime. You could switch this to a request scoped stream info object, but a simpler way would be to just store this on the filter as a member variable. A filter instance is scoped to the request, so this provides a simpler way to carrying over state from the request to the response, without having to store data in stream info (which is more commonly used to share data between different filters).

I think you could also then get rid of a queue, since you're guaranteed a 1-1 relationship between request headers and response headers.